File: app/src/main/main/java/com/timeback/ui/feature/home/HomeViewModel.java
package com.timeback.ui.feature.home;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.timeback.ui.domain.gateway.FeatureGateway;
import com.timeback.ui.domain.model.ScreenState;

import javax.inject.Inject;
import dagger.hilt.android.lifecycle.HiltViewModel;

/**
 * UI-02 홈 대시보드 ViewModel.
 * 오늘의 낭비 시간, 기준선 상태, 목표 진행률을 표시한다.
 */
@HiltViewModel
public class HomeViewModel extends ViewModel {

    private final FeatureGateway gateway;
    private final MutableLiveData<ScreenState> _state = new MutableLiveData<>(new ScreenState.Initial());
    public final LiveData<ScreenState> state = _state;

    @Inject
    public HomeViewModel(FeatureGateway gateway) {
        this.gateway = gateway;
        loadHome();
    }

    public void loadHome() {
        _state.setValue(new ScreenState.Loading());
        gateway.readHome().thenAccept(_state::postValue);
    }

    public void refresh() {
        gateway.refreshHome().thenAccept(_state::postValue);
    }
}

File: src/main/java/com/timeback/ui/feature/home/HomeViewModel.kt
package com.timeback.ui.feature.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.timeback.ui.domain.gateway.FeatureGateway
import com.timeback.ui.domain.model.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * UI-02 홈 대시보드 ViewModel
 * CT-04 §5: ReadHome, RefreshHome
 */
@HiltViewModel
class HomeViewModel @Inject constructor(
    private val gateway: FeatureGateway
) : ViewModel() {

    private val _state = MutableStateFlow<ScreenState>(ScreenState.Initial)
    val state: StateFlow<ScreenState> = _state.asStateFlow()

    init {
        loadHome()
    }

    fun loadHome() {
        viewModelScope.launch {
            _state.value = ScreenState.Loading
            _state.value = gateway.readHome()
        }
    }

    fun refreshHome() {
        viewModelScope.launch {
            val current = _state.value
            if (current is ScreenState.Content<*>) {
                _state.value = ScreenState.Refreshing(current.data)
            }
            _state.value = gateway.refreshHome()
        }
    }
}
